spec: overview of the p2p implementation in v0.34 (#9126)
tendermint/tendermint: spec/p2p/v0.34/pex-protocol.md

# Peer Exchange Protocol

The Peer Exchange (PEX) protocol enables nodes to exchange peer addresses, thus
implementing a peer discovery mechanism.

The PEX protocol uses two messages:

- `PexRequest`: sent by a node to [request](#requesting-addresses) peer
  addresses to a peer
- `PexAddrs`: a list of peer addresses [provided](#providing-addresses) to a
  peer as response to a `PexRequest` message

While all nodes, with few exceptions, participate on the PEX protocol,
a subset of nodes, configured as [seed nodes](#seed-nodes) have a particular
role in the protocol.
They crawl the network, connecting to random peers, in order to learn as many
peer addresses as possible to provide to other nodes.

## Requesting Addresses

A node requests peer addresses by sending a `PexRequest` message to a peer.

For regular nodes, not operating in seed mode, a PEX request is sent when
the node *needs* peers addresses, a condition checked:

1. When an *outbound* peer is added, causing the node to request peer addresses

>    to the new peer

Reviewer: I would rephrase this as " request addresses from the new peer"

Author: Adopted.

> 1. Periodically, by the `ensurePeersRoutine`, causing the node to request peer

Reviewer 2: `1.` -> `2.`

Reviewer: Thank you for taking the time to review the spec. We have merged this PR because we are preparing a next one with new edits and will then open a PR against `main`. I will incorporated your suggestions there though.

[less than 1000 records](./addressbook.md#need-for-addresses).
It is thus reasonable to assume that the common case is that a peer needs more
peer addresses, so that PEX requests are sent whenever the above two situations happen.

A PEX request is sent when a new *outbound* peer is added.
The same does not happen with new inbound peers
because outbound peers, that the node has dialed, are considered more trustworthy
than inbound peers, that the node has accepted.
Moreover, when a node is short of peer addresses, it dials the configured seed nodes;
since they are added as outbound peers, the node can immediately request peer addresses.

The `ensurePeersRoutine` periodically checks, by default every 30 seconds (`ensurePeersPeriod`),
whether the node has enough outbound peers.
If it does not have, the node tries dialing some peer addresses stored in the address book.
As part of this procedure, the node selects a peer at random,
from the set of connected peers retrieved from the switch,
and sends a PEX request to the selected peer.

Sending a PEX request to a peer is implemented by the `RequestAddrs` method of
the PEX reactor.

### Responses

> After sending a PEX request is sent to a peer, the node expect to receive,

Reviewer: ```suggestion
After a PEX request is sent to a peer, the node expects to receive,
```

Author: Manually adopted.

as a response, a `PexAddrs` message from the peer.
This message encodes a list of peer addresses that are

> [added to address book](./addressbook.md#new-addresses),

Reviewer: ```suggestion
[added to the address book](./addressbook.md#new-addresses),
```

having the peer from which the PEX response was received as their source.

Received PEX responses are handled by the `ReceiveAddrs` method of the PEX reactor.

> In the case that of a PEX response received from a peer which is configured as

Reviewer: ```suggestion
In the case of a PEX response received from a peer which is configured as
```

a seed node, the PEX reactor attempts immediately to dial the provided peer
addresses, as detailed [here](TODO).

### Misbehavior

Sending multiple PEX requests to a peer, before receiving a reply from it,
is considered a misbehavior.
To prevent it, the node maintains a `requestsSent` set of outstanding
requests, indexed by destination peers.
While a peer ID is present in the `requestsSent` set, the node does not send
further PEX requests to that peer.
A peer ID is removed from the `requestsSent` set when a PEX response is
received from it.

Sending a PEX response to a peer that has not requested peer addresses
is also considered a misbehavior.
So, if a PEX response is received from a peer that is not registered in
the `requestsSent` set, a `ErrUnsolicitedList` error is produced.
This leads the peer to be disconnected and marked as a bad peer.

## Providing Addresses

When a node receives a `PexRequest` message from a peer,

> it replies to peer with a `PexAddrs` message.

Reviewer: ```suggestion
it replies with a `PexAddrs` message.
```

A seed node is a node configured to operate in `SeedMode`.

### Crawling peers

Seed nodes crawl the network, connecting to random peers and sending PEX
requests to them, in order to learn as many peer addresses as possible.
More specifically, a node operating in seed mode sends PEX requests in two cases:

1. When a outbound peer is added, and the seed node needs more peer addresses,
   it requests peer addresses to the new peer
1. Periodically, the `crawlPeersRoutine` sends PEX requests to a random set of
   peers, whose addresses are registered in the Address Book

The first case also applies for nodes not operating in seed mode.
The second case replaces the second for regular nodes, as seed nodes do not
run the `ensurePeersRoutine`, as regular nodes,
but run the `crawlPeersRoutine`, which is not run by regular nodes.

The `crawlPeersRoutine` periodically, every 30 seconds (`crawlPeerPeriod`),
starts a new peer discovery round.
First, the seed node retrieves a random selection of peer addresses from its
Address Book.
This selection is produced in the same way as in the random selection of peer
addresses that are [provided](#providing-addresses) to a requesting peer.

> Are removed from this selection peers that the seed node has crawled recently,

Reviewer: ```suggestion
From this selection peers that the seed node has crawled recently are removed,
```

Author: I adopted the suggestion, but with a different format.

> last than 2 minutes ago (`minTimeBetweenCrawls`).

Reviewer: ```suggestion
less than 2 minutes ago (`minTimeBetweenCrawls`).
```

a peer in the multiple Tendermint protocols, but because they believe that a
seed node is a good source of addresses of nodes to which they can establish
connections and interact in the multiple Tendermint protocols.

So, when a seed node receives a `PexRequest` message from an inbound peer,
it sends a `PexAddrs` message, containing a selection of peer
addresses, back to the peer and *disconnects* from it.
Seed nodes therefore treat inbound connections from peers as a short-term
connections, exclusively intended to retrieve peer addresses.
Once the requested peer addresses are sent, the connection with the peer is closed.

Moreover, the selection of peer addresses provided to inbound peers by a seed
node, although still essentially random, has a [bias toward old
addresses](./addressbook.md#random-selection-with-bias).
The selection bias is defined by `biasToSelectNewPeers`, hard-coded to `30%`,
meaning that `70%` of the peer addresses provided by a seed node are expected
to be old addresses.
Although this nomenclature is not clear, *old* addresses are the addresses that
survived the most in the address book, that is, are addresses that the seed

> node believe that belong to *good* peers (more details [here](./addressbook.md#good-peers)).

Reviewer: ```suggestion
node believes belong to *good* peers (more details [here](./addressbook.md#good-peers)).
```

Another distinction is on the handling of potential [misbehavior](#misbehavior-1)
of peers requesting addresses.
A seed node does not enforce, a priori, a minimal interval between PEX requests
from inbound peers.
Instead, it does not reply to more than one PEX request per peer inbound
connection, and, as above mentioned, it disconnects from incoming peers after
responding to them.
If the same peer dials again to the seed node and requests peer addresses, the
seed node will reply to this peer like it was the first time it has requested
peer addresses.

> This is more an implementation restriction than a desired behavior.
> The `lastReceivedRequests` map stores the last time a PEX request was
> received from a peer, and the entry relative to a peer is removed from this
> map when the peer is disconnected.
>

> > It is debatable whether this approach indeed prevent abuse against seed nodes.

Reviewer: While we should be concerned with abusing a seed node, we should also consider the scenario where in a network with few seed nodes a peer might be right to request addresses more then once. Maybe we can have an algorithm where we provide to the peer more then one address? Or each peer sends a desired number of addresses (this number can be derived from the currently connected peers and the desired number of peers)?
